src/engine/ChunkSystemTileMap.ts

```diff
@@ -0,0 +1,466 @@
+  public readonly rows: number;
+  public readonly chunkCols: number;
+  public readonly chunkRows: number;
+  public readonly chunkGenerator: ChunkGenerator;
+  public readonly chunkGarbageCollectorPredicate: ChunkSystemGarbageCollectorPredicate;
+  private readonly _chunks: Array<Array<CachedTileMap | undefined> | undefined>;
+  private _chunksXOffset: number;
+  private _chunksYOffset: number;
+  private readonly _chunksToRender: CachedTileMap[];
+  private readonly _spriteSheets: { [key: string]: SpriteSheet };
+  private readonly _chunkRenderingCachePredicate: null | ChunkRenderingCachePredicate;
+
+  constructor(config: ChunkSystemTileMapArgs) {
+    if (config.chunkSize <= 0 || !isSafeInteger(config.chunkSize)) {
+      throw new TypeError(`The chunkSize option must be a positive integer, ${config.chunkSize} was provided`);
+    }
+    if (config.rows <= 0 || !isSafeInteger(config.rows)) {
+      throw new TypeError(`The rows option must be a positive integer, ${config.rows} was provided`);
+    }
+    if (config.cols <= 0 || !isSafeInteger(config.cols)) {
+      throw new TypeError(`The cols option must be a positive integer, ${config.cols} was provided`);
+    }
+    if (config.cols % config.chunkSize) {
+      throw new Error(
+        `The cols option must be a multiple of the chunkSize option, ${config.cols} was provided for the cols option, ${config.chunkSize}` +
+          ' was provided for the chunkSize option.'
+      );
+    }
+    if (config.rows % config.chunkSize) {
+      throw new Error(
+        `The rows option must be a multiple of the chunkSize option, ${config.rows} was provided for the rows option, ${config.chunkSize}` +
+          ' was provided for the chunkSize option.'
+      );
+    }
+
+    super();
+
+    this.x = config.x;
+    this.y = config.y;
+    this.cellWidth = config.cellWidth;
+    this.cellHeight = config.cellHeight;
+    this.chunkSize = config.chunkSize;
+    this.cols = config.cols;
+    this.rows = config.rows;
+    this.chunkCols = this.cols / this.chunkSize;
+    this.chunkRows = this.rows / this.chunkSize;
+    this.chunkGenerator = config.chunkGenerator;
+    this.chunkGarbageCollectorPredicate = config.chunkGarbageCollectorPredicate;
+    this._chunks = [];
+    this._chunksXOffset = 0;
+    this._chunksYOffset = 0;
+    this._chunksToRender = [];
+    this._spriteSheets = {};
+    this._chunkRenderingCachePredicate = config.chunkRenderingCachePredicate || null;
+  }
+
+  public registerSpriteSheet(key: string, spriteSheet: SpriteSheet): void {
+    this._spriteSheets[key] = spriteSheet;
+    for (let rowIndex = 0; rowIndex < this._chunks.length; rowIndex++) {
+      const chunkRow = this._chunks[rowIndex];
+      for (let columnIndex = 0; columnIndex < chunkRow.length; columnIndex++) {
+        const chunk = chunkRow[columnIndex];
+        if (chunk) {
+          chunk.registerSpriteSheet(key, spriteSheet);
+        }
+      }
+    }
+  }
+
+  public collides(actor: Actor): Vector | null {
```

Comment:
I have moved the ChunkSystemTileMap's implementation, and deprecated TileMap.collides in favor of the new shared code. Since there are minor cosmetic differences, I think it to be safest to preserve the current implementation of TileMap.collides for a while.